New function: distributed compute point locations (and tests)

GivAlz opened the pull request:
**Working version with tests**: now reviews are needed :)

It took me a while because there was a small bug with indices which resulted in a small number of wrong points in certain cases: it was difficult to understand where exactly the problem was.

The function currently asks for: cache, a vector of bounding boxes (the one describing the locally owned part of the mesh, so that its creation is up to the user) and a vector of points.

The mpi communicator is recovered from the cache's triangulation: I used dynamic casting and an assertion error is thrown if no "parallel::Triangulation" or derived object is used.

There are two tests:
1. Compares the result with the serial version of the code.
2. The points are taken from a shared triangulation: this means they can be computed "by hand" and then compared with the distributed compute point locations result.

The function's code is quite articulated: I tried to comment and organize as much as possible, but if some parts are unclear/too ugly please tell me and I shall try to improve them.

luca-heltai commented on include/deal.II/distributed/grid_tools.h:
```diff
@@ -50,6 +50,43 @@ namespace parallel
+        std::vector< std::vector< Point<spacedim> > >,
+        std::vector< unsigned int >
+        >
+        distributed_compute_point_locations
```
Since `distributed` is already in the namespace, I would just use the same name of the other function: `compute_point_locations`.

tjhei commented:
A couple of questions/thoughts:
- The method will return only the subset of all points that are in the locally owned area, right?
- Do you want to guarantee that each Point will be returned on exactly one processor? This might be a useful condition, but probably difficult to implement efficiently.
- I wonder if you can change the method to only do the communication part and leave the cell and unit coordinate computation to a separate call to ``compute_point_locations``. That would make the method smaller and more flexible. Even better would be if you could ship any additional data around (particles probably have other data attached to it).

GivAlz commented:
Hi! I've started writing the function (I couldn't debug it/add a test though...especially because I shall need the function to compute global bounding boxes and it's cumbersome).
Anyway here are some answers @tjhei : 
- Yes, the method return points which are inside the locally owned area (contained in get<3>(output tuple) )
- Yes, I guarantee the point are given as output from a single process: what I'm currently doing is running over all cells and check if the cell is locally owned or not: the points are added to output only in the first case
- Currently I use compute_point_locations also to understand in which cell the point is owned and, moreover, particles need the reference location to be created. But, especially because the function is so huge, I guess it would be a good idea to add a function which uses bounding boxes to send "things around" but which does only guarantee that also "the process with the locally owned part of the mesh gets the particle/whatever". Then add this function (for the point case) which runs compute point locations and throws away non-local points.

@luca-heltai  I modified cache adding a new value 0x010 (I hope I understood how the things work) but, clearly, it's currently missing a function to update the value.
Also what I'm currently doing with the output is trying to (needs to be debugged) guarantee that the output tuple(cell/qpoints/maps/points/rank) work similarly to the serial case (i.e. points(maps[c][pt]) is the transformed of qpoints[c] point pt). Because I need to do a lot of things (such as trowing away points which are in artificial&ghost cells) I don't think we can use maps as a local_id for the point...

GivAlz commented:
Update: ok, still not finished, but here's the first part of the function using the various tools created in these days...tomorrow I shall finish and hopefully get a first working version.

@luca-heltai suggested me to use threads, but it seems to work only for single function calls so I hope the way I'm doing it is reasonable...

luca-heltai commented:
Please rebase and fix conflicts

GivAlz commented:
@luca-heltai before merging I need also to update this version to the better-scaling one which uses unordered maps and add another test, working on it :)

GivAlz commented:
I just pushed a working version; the best so far which scales reasonably well with a few dozens of cores. Before merging I think at least three things should be done:
- currently the "global bounding box description" is set with the parameters I personally used; this clearly isn't a general solution. I shall add parameters to distributed compute point locations, which is ugly but at least more general (but in the future Cache should handle this part)
- currently there is one other test, I need to perfect another one (which uses vertices of a shared mesh as points: this allows to check with compute point locations if everything is working properly)
- probably the documentation can be improved

This overall shouldn't take long. The function is quite complex so I'm sure there's plenty to be improved: if there are comments and suggestions I'm here!

Once this PR is finished I shall work again on compute mesh predicate (I'm searching in literature for better algorithms) and will probably modify the structure so that it uses a tree of bounding boxes which should scale well also with many cores; but this is a matter for the future.

Thanks for reading.
Best,
Giovanni

bangerth commented:
@GivAlz -- it's not clear to me from your description whether the current state of this pull request is "finished and ready for a code review", or "this is still work in progress which I'll put here in case someone wants to look at it already". 

I think it would be useful if you made this clear in your descriptions of pull requests, in particular if it is of the latter kind -- in that case, please say so explicitly, for example in the first sentence of a new post. Right now, the list of comments above reads a lot like "Here is a first attempt at something, but it's not finished...", "I did some more work on this, and it looks better, but not quite right yet...". This makes it rather difficult to know what the current state is, and what you hope others should do at a given moment with this pull request.

GivAlz commented:
Still WIP: do not merge but look at it if you want to.

@bangerth  You're right, sorry; this later days have been intense. I shall make some additions soon (adding a test and a small part) and then ask for it to be fully reviewed!

luca-heltai commented:
Can you edit the first comment, and make a checkmarked list with what you have done and what needs to be done? It is very difficult now to figure out what's going on...

GivAlz commented:
Hi! I just updated the code and the first comment; what I did was:
- make the function take as input the description of its locally owned cells using bounding boxes (this way the options can be handled by the user and, in the future, cache->local description or similar shall be passed)
- added two test: one compares the function with its serial counterpart, the other tests with a distributed triangulation

The mpi communicator still needs to be given as an option ( get_communicator () works only for parallel::Triangulation...I shall modify this).

The other problem is that there still are some small bugs in the code. For the first kind of bugs I don't know how to tackle them properly: test 02 returns the following error when called using "ctest" with the debug version:

> 2066: mpiexec detected that one or more processes exited with non-zero status, thus causing
> 2066: the job to be terminated. The first process to do so was:
> 2066: 
> 2066:   Process name: [[29945,1],2]
> 2066:   Exit code:    127

while if called with the following command:

> mpirun -np 2 testname.debug

it works perfectly (i.e. no error is thrown and test pass)...if someone knows why there is this difference in behaviour, please tell me.

The other bug is more subtle: while the function works, in general, with some configuration of "create local description of the mesh" and certain number of processes the test 02 fails, but I still have to figure out what exactly goes wrong (and I think I shall add a test for these cases...which might be in the other function, I shall update soon enough on this).

luca-heltai commented:
You can dynamic cast a Triangulation to a p::d::triangulation. If the cast succeeds, then you can query the pointer for the communicator. If not, you throw an exception, and say that a p::d::tria is required. 

In fact, this should also work for a parallel::shared::tria. Is there a reason why you are not allowing it in that case?

GivAlz commented:
@luca-heltai Now it works with any parallel::Triangulation (thank you for suggesting dynamic casting!) but if the triangulation is shared then it's probably better to use compute point locations and then just send around the results to the owners.

@tjhei Thank you again for the suggestion you sent: this is still a "basic" version which works well with a few dozen processes (maybe more, I shall test this), as the search is done one Bounding Box at a time. Once it is perfected, it shouldn't be to difficult to use some library to handle AABB trees and obtain good scalability.

Thank you everyone for reading and for your suggestions: I know the function is a bit complicated so I hope it won't take too long reviewing...I tried polishing it as much as possible.

All the best,
Giovanni

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -661,6 +661,58 @@ namespace GridTools
+   * Distributed compute point locations is a function similar to
+   * GridTools::compute_point_locations but working for parallel::Triangulation
+   * objects and, unlike its serial version, also for a distributed triangulation
+   * ( see parallel::distributed::Triangulation ).
```
please remove empty spaces here

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -661,6 +661,58 @@ namespace GridTools
+   * GridTools::compute_point_locations but working for parallel::Triangulation
+   * objects and, unlike its serial version, also for a distributed triangulation
+   * ( see parallel::distributed::Triangulation ).
+   * Given a @p cache and a list of
```
Maybe start with this and then mention the test in the above (that it's similar to `GridTools::compute_point_locations`)

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -661,6 +661,58 @@ namespace GridTools
+   * @param[in] local_points the array of points owned by the current process. Every
+   *  process can have a different array of points which can be empty and not
+   *  contained within the locally owned part of the triangulation
+   * @param[in] the description of the locally owned part of the mesh made with bounding
```
`local_bbox` is missing here, i guess.

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -661,6 +661,58 @@ namespace GridTools
+   * @param[out] tuple containing the quadrature information
+   *
+   * The elements of the output tuple are:
+   * - cells : a vector of a vector cells of the all cells containing at
```
a vector of ~~a vector cells of the all~~ cells containing at least a single point. (?)

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -661,6 +661,58 @@ namespace GridTools
+   * The elements of the output tuple are:
+   * - cells : a vector of a vector cells of the all cells containing at
+   *  least a point.
+   * - qpoints : a vector of vector of points. Each entry contains the locally owned
```
"Each entry contains a vector locally owned points positions in coordinates of the unit cell" (?)

I am not a native speaker, but "points transformed in the unit cell" sounded strange to me.

GivAlz replied:
Yes, it sounds awful...I shall copy and paste from compute point locations where, I see, someone with better English skills did a great job!

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -661,6 +661,58 @@ namespace GridTools
+   *  least a point.
+   * - qpoints : a vector of vector of points. Each entry contains the locally owned
+   *  points transformed in the unit cell.
+   * - maps : a vector indices of vector of integers, containing the mapping between
```
a vector ~~indices~~ of vector of integers

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -661,6 +661,58 @@ namespace GridTools
+   * - qpoints : a vector of vector of points. Each entry contains the locally owned
+   *  points transformed in the unit cell.
+   * - maps : a vector indices of vector of integers, containing the mapping between
+   *  the numbering in qpoints, and the vector of local points of the process owning
```
..in qpoints (previous element of the tuple),...

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -661,6 +661,58 @@ namespace GridTools
+   *  points transformed in the unit cell.
+   * - maps : a vector indices of vector of integers, containing the mapping between
+   *  the numbering in qpoints, and the vector of local points of the process owning
+   *  the points.
```
without looking into the code, i did not understand what "the vector of local points of the process owning" means. Myabe 

> the input vector @p local_points

 ?

GivAlz replied:
Not really... some of these points might be "owned" by another process and then communicated to the local one...so, usually, some are copies of local points, some are new... I'll try to explain this better

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -661,6 +661,58 @@ namespace GridTools
+   * - maps : a vector indices of vector of integers, containing the mapping between
+   *  the numbering in qpoints, and the vector of local points of the process owning
+   *  the points.
+   * - points : a vector of vector of points. points[i][j] is the real point corresponding
```
what is "real point" ?

GivAlz replied:
If I write: 
@p points[i][j] is the point in the real space corresponding to @p qpoints[i][j] 

it is more understandable? I just want to separate between "points on the unit cell" (qpoints) and "points in the real space, where the mesh is" (points)

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -661,6 +661,58 @@ namespace GridTools
+   * - points : a vector of vector of points. points[i][j] is the real point corresponding
+   *  to qpoints[i][j]
+   * - owners : a vector of vectors containing the rank of the process owning the
+   *  corresponding element of points.
```
maybe: owning the corresponding point (the previous element of the tuple) ?

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -661,6 +661,58 @@ namespace GridTools
+   *
+   * The function uses the triangulation's mpi communicator: for this reason it
+   * throws an assert error if the Triangulation is not derived from
+   * parallel::Triangulation .
```
maybe just use MPI_COMM_SELF ?

GivAlz replied:
Do you mean to make the function work with any sort of triangulation and use MPI_COMM_SELF if the triangulation is not parallel? I didn't think of this but it should work!

davydden commented on include/deal.II/grid/grid_tools.h:
```diff
@@ -661,6 +661,58 @@ namespace GridTools
+   *  the points.
+   * - points : a vector of vector of points. points[i][j] is the real point corresponding
+   *  to qpoints[i][j]
+   * - owners : a vector of vectors containing the rank of the process owning the
```
btw, probably enumerated list fits better `std::tuple` description as you access its elements via `std::get<0>(...)` and alike. Downside is that enumeration in doxygen will probably starts with 1 instead of 0.

GivAlz replied:
I think it will start with 1, yes, and the solution seems a bit impractical to me (https://stackoverflow.com/questions/18208906/start-doxygen-list-at-zero)...maybe just adding the number after the "-"? Such as
- 0. cells:...
I know it's a bit ugly, but at least it clearly states the connection std::get<0> to cells..

masterleinad commented:
/run-tests

GivAlz commented:
@davydden Thank you for all the comments: it is very helpful having someone checking line by line. I tried to do my best, the only remaining part might be allowing for a general triangulation and using MPI_COMM_SELF if the triangulation is not parallel... what do you think @luca-heltai 

There was a small error for the serial case, it now should compile properly.

luca-heltai commented:
There seems to be a problem with the test in serial:

~~~
===============================   OUTPUT BEGIN  ===============================
grid/distributed_compute_point_locations_01.debug: BUILD failed. Output:
[100%] Generating distributed_compute_point_locations_01.debug/interrupt_guard.cc
Scanning dependencies of target distributed_compute_point_locations_01.debug
Building CXX object CMakeFiles/distributed_compute_point_locations_01.debug.dir/distributed_compute_point_locations_01.cc.o
/home/bob/source/tests/grid/distributed_compute_point_locations_01.cc: In instantiation of â€˜void test_compute_pt_loc(unsigned int) [with int dim = 2]â€™:
/home/bob/source/tests/grid/distributed_compute_point_locations_01.cc:130:29:   required from here
/home/bob/source/tests/grid/distributed_compute_point_locations_01.cc:40:66: error: no matching function for call to â€˜dealii::parallel::distributed::Triangulation<2, 2>::Triangulation(MPI_Comm&)â€™
   parallel::distributed::Triangulation<dim> tria(mpi_communicator);
                                                                  ^
/home/bob/source/tests/grid/distributed_compute_point_locations_01.cc:40:66: note: candidates are:
In file included from /home/bob/source/tests/grid/distributed_compute_point_locations_01.cc:26:0:
/home/bob/source/include/deal.II/distributed/tria.h:1030:7: note: dealii::parallel::distributed::Triangulation<<anonymous>, <anonymous> >::Triangulation() [with int dim = 2; int spacedim = 2] <deleted>
       Triangulation () = delete;
       ^
/home/bob/source/include/deal.II/distributed/tria.h:1030:7: note:   candidate expects 0 arguments, 1 provided
/home/bob/source/include/deal.II/distributed/tria.h:1023:11: note: dealii::parallel::distributed::Triangulation<2, 2>::Triangulation(const dealii::parallel::distributed::Triangulation<2, 2>&) <deleted>
     class Triangulation : public dealii::parallel::Triangulation<dim,spacedim>
           ^
/home/bob/source/include/deal.II/distributed/tria.h:1023:11: note:   no known conversion for argument 1 from â€˜MPI_Comm {aka int}â€™ to â€˜const dealii::parallel::distributed::Triangulation<2, 2>&â€™
/home/bob/source/include/deal.II/distributed/tria.h:1023:11: note: dealii::parallel::distributed::Triangulation<2, 2>::Triangulation(dealii::parallel::distributed::Triangulation<2, 2>&&) <deleted>
/home/bob/source/include/deal.II/distributed/tria.h:1023:11: note:   no known conversion for argument 1 from â€˜MPI_Comm {aka int}â€™ to â€˜dealii::parallel::distributed::Triangulation<2, 2>&&â€™
/home/bob/source/tests/grid/distributed_compute_point_locations_01.cc:57:59: error: no matching function for call to â€˜compute_mesh_predicate_bounding_box(const dealii::Triangulation<2, 2>&, dealii::IteratorFilters::LocallyOwnedCell&, int, bool, int)â€™
                                                 1, true, 4); // These options should be passed
                                                           ^
/home/bob/source/tests/grid/distributed_compute_point_locations_01.cc:57:59: note: candidate is:
In file included from /home/bob/source/tests/grid/distributed_compute_point_locations_01.cc:24:0:
/home/bob/source/include/deal.II/grid/grid_tools.h:1345:3: note: template<class MeshType> std::vector<dealii::BoundingBox<MeshType:: space_dimension> > dealii::GridTools::compute_mesh_predicate_bounding_box(const MeshType&, const std::function<bool(const typename MeshType::active_cell_iterator&)>&, const unsigned int&, const bool&, const unsigned int&)
   compute_mesh_predicate_bounding_box
   ^
/home/bob/source/include/deal.II/grid/grid_tools.h:1345:3: note:   template argument deduction/substitution failed:
/home/bob/source/tests/grid/distributed_compute_point_locations_01.cc:57:59: note:   â€˜dealii::IteratorFilters::LocallyOwnedCellâ€™ is not derived from â€˜const std::function<bool(const typename MeshType::active_cell_iterator&)>â€™
                                                 1, true, 4); // These options should be passed
                                                           ^
/home/bob/source/tests/grid/distributed_compute_point_locations_01.cc: In instantiation of â€˜void test_compute_pt_loc(unsigned int) [with int dim = 3]â€™:
/home/bob/source/tests/grid/distributed_compute_point_locations_01.cc:131:29:   required from here
/home/bob/source/tests/grid/distributed_compute_point_locations_01.cc:40:66: error: no matching function for call to â€˜dealii::parallel::distributed::Triangulation<3, 3>::Triangulation(MPI_Comm&)â€™
   parallel::distributed::Triangulation<dim> tria(mpi_communicator);
                                                                  ^
/home/bob/source/tests/grid/distributed_compute_point_locations_01.cc:40:66: note: candidates are:
In file included from /home/bob/source/tests/grid/distributed_compute_point_locations_01.cc:26:0:
/home/bob/source/include/deal.II/distributed/tria.h:1030:7: note: dealii::parallel::distributed::Triangulation<<anonymous>, <anonymous> >::Triangulation() [with int dim = 3; int spacedim = 3] <deleted>
       Triangulation () = delete;
       ^
/home/bob/source/include/deal.II/distributed/tria.h:1030:7: note:   candidate expects 0 arguments, 1 provided
/home/bob/source/include/deal.II/distributed/tria.h:1023:11: note: dealii::parallel::distributed::Triangulation<3, 3>::Triangulation(const dealii::parallel::distributed::Triangulation<3, 3>&) <deleted>
     class Triangulation : public dealii::parallel::Triangulation<dim,spacedim>
           ^
~~~

masterleinad commented:
> There seems to be a problem with the test in serial:

`parallel::distributed::Triangulation` is just not implemented properly if we compiler without MPI support.
Did you intend to use a regular `dealii::Triangulation` object here?

davydden commented:
> There seems to be a problem with the test in serial:

or just rename the output file to have `mpirun=1` to request MPI.

luca-heltai commented:
@GivAlz , can you rename the test and fix the conflicts?

GivAlz commented:
Ok, I did not merge correctly, now it compiles and passes tests. In the end I used mpirun=1, thank you everyone!!!

luca-heltai commented:
@davydden , Merge?

davydden commented:
> @davydden , Merge?

I approved already, but I would still squash.

GivAlz commented:
Squashed :)

jppelteret commented:
@GivAlz This appears to have failed on the one tester :-(

```
[647/812] Building CXX object source/grid/CMakeFiles/obj_grid_debug.dir/grid_tools.cc.o
FAILED: source/grid/CMakeFiles/obj_grid_debug.dir/grid_tools.cc.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DDEBUG -DTBB_DO_ASSERT=1 -DTBB_USE_DEBUG -Isource/grid -Iinclude -I/Users/cib-osx/auto-dealii-tjhei-osx/dealii/include -I/Users/cib-osx/auto-dealii-tjhei-osx/dealii/bundled/tbb41_20130401oss/include -I/Users/cib-osx/auto-dealii-tjhei-osx/dealii/bundled/umfpack/UMFPACK/Include -I/Users/cib-osx/auto-dealii-tjhei-osx/dealii/bundled/umfpack/AMD/Include -I/Users/cib-osx/auto-dealii-tjhei-osx/dealii/bundled/boost-1.62.0/include -I/Users/cib-osx/auto-dealii-tjhei-osx/dealii/bundled/muparser_v2_2_4/include -pedantic -fPIC -Wall -Wextra -Wpointer-arith -Wwrite-strings -Wsynth -Wsign-compare -Wswitch -Woverloaded-virtual -Wno-deprecated-declarations -Qunused-arguments -Wno-unsupported-friend -Wno-undefined-var-template -openmp-simd -std=c++11 -ftemplate-depth=1024 -Werror -Wno-parentheses -Wno-unused-local-typedefs -O0 -ggdb -Wa,--compress-debug-sections -MD -MT source/grid/CMakeFiles/obj_grid_debug.dir/grid_tools.cc.o -MF source/grid/CMakeFiles/obj_grid_debug.dir/grid_tools.cc.o.d -o source/grid/CMakeFiles/obj_grid_debug.dir/grid_tools.cc.o -c /Users/cib-osx/auto-dealii-tjhei-osx/dealii/source/grid/grid_tools.cc
/Users/cib-osx/auto-dealii-tjhei-osx/dealii/source/grid/grid_tools.cc:4357:3: error: control reaches end of non-void function [-Werror,-Wreturn-type]
  }
  ^
source/grid/grid_tools.inst:672:2: note: in instantiation of function template specialization 'dealii::GridTools::distributed_compute_point_locations<1, 1>' requested here
 distributed_compute_point_locations
 ^
/Users/cib-osx/auto-dealii-tjhei-osx/dealii/source/grid/grid_tools.cc:4357:3: error: control reaches end of non-void function [-Werror,-Wreturn-type]
  }
  ^
source/grid/grid_tools.inst:723:2: note: in instantiation of function template specialization 'dealii::GridTools::distributed_compute_point_locations<1, 2>' requested here
 distributed_compute_point_locations
 ^
/Users/cib-osx/auto-dealii-tjhei-osx/dealii/source/grid/grid_tools.cc:4357:3: error: control reaches end of non-void function [-Werror,-Wreturn-type]
  }
  ^
source/grid/grid_tools.inst:774:2: note: in instantiation of function template specialization 'dealii::GridTools::distributed_compute_point_locations<1, 3>' requested here
 distributed_compute_point_locations
 ^
/Users/cib-osx/auto-dealii-tjhei-osx/dealii/source/grid/grid_tools.cc:4357:3: error: control reaches end of non-void function [-Werror,-Wreturn-type]
  }
  ^
source/grid/grid_tools.inst:876:2: note: in instantiation of function template specialization 'dealii::GridTools::distributed_compute_point_locations<2, 2>' requested here
 distributed_compute_point_locations
 ^
/Users/cib-osx/auto-dealii-tjhei-osx/dealii/source/grid/grid_tools.cc:4357:3: error: control reaches end of non-void function [-Werror,-Wreturn-type]
  }
  ^
source/grid/grid_tools.inst:927:2: note: in instantiation of function template specialization 'dealii::GridTools::distributed_compute_point_locations<2, 3>' requested here
 distributed_compute_point_locations
 ^
/Users/cib-osx/auto-dealii-tjhei-osx/dealii/source/grid/grid_tools.cc:4357:3: error: control reaches end of non-void function [-Werror,-Wreturn-type]
  }
  ^
source/grid/grid_tools.inst:1080:2: note: in instantiation of function template specialization 'dealii::GridTools::distributed_compute_point_locations<3, 3>' requested here
 distributed_compute_point_locations
 ^
6 errors generated.
```

davydden commented on source/grid/grid_tools.cc:
```diff
@@ -3315,6 +3791,572 @@ namespace GridTools
+    (void)cache;
+    (void)local_points;
+    (void)local_bbox;
+    Assert(false, ExcMessage("GridTools::distributed_compute_point_locations() requires MPI."));
```
still need to return something here.

luca-heltai commented:
@GivAlz , as suggested by @davydden , you need to add something like:

~~~
--- a/source/grid/grid_tools.cc
+++ b/source/grid/grid_tools.cc
@@ -4156,6 +4156,14 @@ next_cell:
     (void)local_points;
     (void)local_bbox;
     Assert(false, ExcMessage("GridTools::distributed_compute_point_locations() requires MPI."));
+    std::tuple<
+    std::vector< typename Triangulation<dim, spacedim>::active_cell_iterator >,
+        std::vector< std::vector< Point<dim> > >,
+        std::vector< std::vector< unsigned int > >,
+        std::vector< std::vector< Point<spacedim> > >,
+        std::vector< std::vector< unsigned int > >
+        > tup;
+    return tup;
 #else
     // Recovering the mpi communicator used to create the triangulation
     const auto &tria_mpi =
~~~
for compilation to go through without warnings (which are now considered as errors).

GivAlz commented:
Thank you @luca-heltai  and @davydden .
Sorry everyone if this took quite long but these last ten days were quite busy and strange for unavoidable personal reasons, as Luca knows well. Hopefully I shall be now back on track with things everything :)
Best,
Giovanni
